### PR TITLE
fix tb_client eviction race

### PR DESCRIPTION
**Pull request by brson**
This fixes a data race-induced crash in tb_client upon eviction. The race always occurs if an eviction is followed by a submit and cannot be worked around, though it doesn't always crash.

## Problem

(This is copied from a quick writeup I did earlier).

Here's my current understanding of the client eviction crash.
This is not the windows crash but may be related.

The three functions most in play here are client_eviction_callback,
vtable_submit_fn, and io_thread.
client_eviction_callback and io_thread run on the io_thread,
vtable_submit_fn runs on any thread.

The semantics that i think are expressed in the code are:
client_eviction_callback shuts down the io thread while
letting client-handle holders / transaction submitters know
"we're evicted. stop submitting", with the expectation
that at this point the client is shut down.
the fundamental issue this needs to solve is that
other threads may be submitting transactions at the same time.

here's the guy that starts the bug:

```
        fn client_eviction_callback(client: *Client, eviction: *const Message.Eviction) void {
            // ... snip ...
            self.eviction_reason = eviction.header.reason;
            self.signal.stop();
        }
```

This function on the io thread is not called under lock and eviction_reason is shared but not atomic.
that's wrong, but it's not the main problem.

Hitting the stop signal wakes up the io loop (which is on the same thread in this case)
and causes it to start its shutdown sequence.
At this point there may be transactions at various stages of submission that can't be stopped - somehow they either need to go all the way to the server, or to be short-circuited and in either case a response given to the client.

now submit, which runs on any thread:

```
        fn vtable_submit_fn(context: *anyopaque, packet_extern: *Packet.Extern) void {
            // ... snip ...

            if (self.eviction_reason == null) {
                // Enqueue the packet and notify the IO thread to process it asynchronously.
                assert(self.signal.status() == .running);
                self.submitted.push(packet);
                self.signal.notify();
            } else {
                // Cancel the packet since we stop the IO thread during eviction.
                assert(self.signal.status() != .running);
                self.packet_cancel(packet);
            }
        }

```

This function is called under lock but it doesn't protect eviction_reason because
the other thread doesn't take the lock.
Even if it was under lock that's still not _the problem_.

Note that thread sanitizer doesn't see the eviction_reason data race, which looks clear to me;
I don't hthink that Zig 0.14's thread sanitizer instrumentation is complete.

This if/else though should raise alarms because the else is a corner
case that has oddly different synchronization requirements
than any other code in the client: normally the io thread
is responsible for calling packet_cancel, but in this case
suddenly packet_cancel is being called on an arbitrary thread.

That's also not the problem.

This entire if-else block interacts catastrophically with the io_thread,
which is already in the process of shutting itself down and deleting some
of the values being called here in both branches (because eviction_callback
called `signal.stop`).

```
        fn io_thread(self: *Context) void {
            while (self.signal.status() != .stopped) {
                self.tick();
                self.io.run_for_ns(constants.tick_ms * std.time.ns_per_ms) catch |err| {
                    log.err("{}: IO.run() failed: {s}", .{
                        self.client_id,
                        @errorName(err),
                    });
                    @panic("IO.run() failed");
                };
            }

            self.cancel_request_inflight();

            while (self.pending.pop()) |packet| {
                packet.assert_phase(.pending);
                self.packet_cancel(packet);
            }

            // The submitted queue is no longer accessible to user threads,
            // so synchronization is not required here.
            while (self.submitted.pop()) |packet| {
                packet.assert_phase(.submitted);
                self.packet_cancel(packet);
            }

            self.io.cancel_all();
            self.signal.deinit();
            self.client.deinit(self.gpa.allocator());
            self.message_pool.deinit(self.gpa.allocator());
            self.io.deinit();
        }
```

Once `stop` is called the io thread loop exits and all that shutdown code
runs, unsynchronized with client threads, one of which is currently
calling `submit`.

The manifestation of this is that one or any of the signal methods gets called by submit in a client thread on a destroyed `signal` value.

I don't have the actual crash output but it usually happens
in this function, called from submit:

```
    pub fn notify(self: *Signal) void {
        // Try to transition from `waiting` to `notified`.
        // If it fails, analyze the current state to determine if a notification is needed.
        var state: @TypeOf(self.event_state.raw) = .waiting;
        while (self.event_state.cmpxchgStrong(
            state,
            .notified,
            .release,
            .acquire,
        )) |state_actual| {
            switch (state_actual) { // XXXXX crash "switch on corrupt value"
                .waiting, .running => state = state_actual, // Try again.
                .notified => return, // Already notified.
                .shutdown => return, // Ignore notifications after shutdown.
            }
        }

        // snip
    }
```

## The fix

The main insight into the fix is that the client context should not attempt to shut down the io thread itself - this creates a "bimodal" situation where a rare corner case has very different parallel/synchronization behavior, with client submission threads being required to handle packet cancellation, and is hard to reason about and test. After the patch the io thread runs after eviction and continues short-circuiting packet_cancel until the client calls the destructor.

**NOTE that the semantics of eviction are that after a client is evicted all subsequent requests report evicted, and the client must notice this and tear down the client and start over.**
This was true before but was possibly not well understood, and the docs I think are misleading and need to be updated prominently.

This is the current doc on the subject:

> The cluster sends a message to notify the evicted session that it has ended. Typically the evicted                
client is no longer active (already terminated), but if it is active, the eviction message causes it              
to self-terminate, bubbling up to the application as an `session evicted` error.                                  

The 'self-terminate' phrase is misleading - the client continues to exist and can be called even after an eviction, and the only solution is to destroy and recreate. This also means that callers that are sharing the client will need to coordinate its destruction while handling continued eviction responses.

It is not easy to produce a test that quickly and reliably causes the crash using only the public API, so I had to add a private testing API to trigger the eviction.

### CODE REVIEW

**Comment by brson** [comment]
I moved this out of draft because these patches are complete and can be merged as is if desired, but I haven't completed the writeup. I'd like to throw a few other patches at the CI but I can do that on another pr.

**Comment by brson** [comment]
I pushed a revert temporarily to get the CI to show the crash.

**Comment by brson** [comment]
Here's the crash under CI. This is interesting because it is not how I see the crash manifest locally... (this says "reached unreachable code", but the one i see is usually "switch on corrupt value").

https://github.com/tigerbeetle/tigerbeetle/actions/runs/19338091930/job/55318769180?pr=3373

```
...

2025-11-13 16:19:56.824Z info(message_bus): 0: set_and_verify_peer connection from client_likely=159099267403814602199232540234891398031
warning(client): 159099267403814602199232540234891398031: on_reply: slow request, request=0 op=19 size=512 register time=593ms
2025-11-13 16:19:56.826Z info(message_bus): 0: set_and_verify_peer connection from client_likely=242777540024847158844045920161156454233
warning(client): 242777540024847158844045920161156454233: on_reply: slow request, request=0 op=8 size=512 register time=591ms
2025-11-13 16:19:56.831Z error(replica): 0N: on_request: ignoring older session
2025-11-13 16:19:56.831Z warning(replica): 0N: sending eviction message to client=242777540024847158844045920161156454233 reason=session_too_low
error(client): 242777540024847158844045920161156454233: session evicted: reason=session_too_low (cluster_release=0.0.1, client_release=0.0.1)
thread 4880 panic: reached unreachable code
Unable to dump stack trace: debug info stripped
2025-11-13 16:19:57.514Z info(main): stdin closed, exiting
2025-11-13 16:19:57.514Z info(main): stdin closed, exiting
error: test failed, to rerun pass `--test tests`

Caused by:
  process didn't exit successfully: `/home/runner/work/tigerbeetle/tigerbeetle/src/clients/rust/target/debug/deps/tests-ddfd2ef414ef6f87` (signal: 6, SIGABRT: process abort signal)
++++

rust ci: 2m57.961s
2025-11-13 16:19:57.540Z error: ExecNonZeroExitStatus
/home/runner/work/tigerbeetle/tigerbeetle/src/shell.zig:569:42: 0x1303f67 in exec_inner (scripts)
        .Exited => |code| if (code != 0) return error.ExecNonZeroExitStatus,
                                         ^
/home/runner/work/tigerbeetle/tigerbeetle/src/shell.zig:398:5: 0x136f5ac in exec__anon_37366 (scripts)
    return exec_inner(shell, argv.slice(), .{});
    ^
/home/runner/work/tigerbeetle/tigerbeetle/src/clients/rust/ci.zig:8:5: 0x1371cad in tests (scripts)
    try shell.exec("cargo test --all", .{});
    ^
/home/runner/work/tigerbeetle/tigerbeetle/src/scripts/ci.zig:68:21: 0x13887d0 in run_tests (scripts)
                    try ci.tests(shell, gpa);
                    ^
/home/runner/work/tigerbeetle/tigerbeetle/src/scripts/ci.zig:41:9: 0x138946d in main (scripts)
        try run_tests(shell, gpa, cli_args.language);
        ^
/home/runner/work/tigerbeetle/tigerbeetle/src/scripts.zig:99:26: 0x1429a8d in main (scripts)
        .ci => |args_ci| try ci.main(shell, gpa, args_ci),
                         ^
error: the following command exited with code 1 (expected exited with code 0):
/home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/o/6de7778510395fd91c695e7ca2baf735/scripts ci --language=rust 
Build Summary: 4/6 steps succeeded; 1 failed
ci transitive failure
+- run scripts failure
error: the following build command failed with exit code 1:
/home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/o/ef286d337bafdc6384d4b24762658687/build /home/runner/work/tigerbeetle/tigerbeetle/zig/zig zig/lib /home/runner/work/tigerbeetle/tigerbeetle /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache /home/runner/.cache/zig --seed 0x3a0a15f9 -Ze845ecb2c477584f ci -- rust
Error: Process completed with exit code 1.
```

File: src/clients/c/tb_client/context.zig [hunk 1]
```diff
@@ -208,6 +219,8 @@ pub fn ContextType(
         eviction_reason: ?vsr.Header.Eviction.Reason,
```
Comment (chaitanyabhandari) [inline src/clients/c/tb_client/context.zig:218]: Could another way to go here be make `eviction_reason` a `std.atomic.Value` to make accesses thread safe? The reason why I ask is that the motivation of
https://github.com/tigerbeetle/tigerbeetle/pull/2923 was to deinitialize the client upon eviction. Also, because in client.zig we specifiy the semantics of eviction to entail that: https://github.com/tigerbeetle/tigerbeetle/blob/f9c4c763962287ec59d5fa4aa112c6a029aae3df/src/vsr/client.zig#L137-L141


That way, all subsequent submitted packets will still see the eviction response, and we could also shut down the io_thread, which in turn deinitializes the client?

File: src/clients/c/tb_client/context.zig [hunk 2]
```diff
@@ -748,11 +772,21 @@ pub fn ContextType(
...
-            // client before it exits (see `io_thread`).
+            // After eviction the io_thread will short-circuit packet_send to packet_cancel.
             self.eviction_reason = eviction.header.reason;
-            self.signal.stop();
+
+            // This is subtle and ugly - client _does not_ send us a completion callback
+            // for evicted requests, so if we don't do something with it we'll deadlock.
+            // Here we reach into the guts of client and grab the packet to cancel it.
+            // Should be restructured in the future to just hand us the packet as an argument
+            // to this callback, or for client to call both callbacks.
```
Comment (chaitanyabhandari) [inline src/clients/c/tb_client/context.zig]: Curious if we could change the interface between the `context.zig` and `client.zig` to make sure the evicted request is handled in a similar way non evicted requests are (i.e. client passes us `raw_user_data` which we use to derive the `Packet`) in a _non breaking fashion_? IIUC, it shouldn't be a breaking change as there's no user-API-level, changes, no?

File: src/clients/c/tb_client/context.zig [hunk 3]
```diff
@@ -748,11 +772,21 @@ pub fn ContextType(
...
+            self.cancel_request_inflight();
+
+            // Cancel every packet.
+            while (self.pending.pop()) |packet| {
+                packet.assert_phase(.pending);
+                self.packet_cancel(packet);
+            }
```
Comment (chaitanyabhandari) [inline src/clients/c/tb_client/context.zig]: Double checking my understanding here about _why_ this is thread safe:

So, this is accessed on the `io_thread`, because its in a callback, and then only place `self.pending` is accessed is _also_ on the io_thread, since its invoked via signal_notify_callback -> packet_enqueue. So, its (thread) safe to do this, as there can be no contention on self.pending.

File: src/clients/c/tb_client/context.zig [hunk 4]
```diff
@@ -322,6 +335,7 @@ pub fn ContextType(
                 .completion_context_fn = &vtable_completion_context_fn,
                 .deinit_fn = &vtable_deinit_fn,
                 .init_parameters_fn = &vtable_init_parameters_fn,
+                .trigger_eviction_for_testing_fn = &vtable_trigger_eviction_for_testing_fn,
```
Comment (chaitanyabhandari) [inline src/clients/c/tb_client/context.zig]: I was thinking of alternative ways to test eviction that could avoid adding this new interface! For that, I exhaustively went through all the eviction reasons:

```C
        pub const Reason = enum(u8) {
            reserved = 0,
            no_session = 1,
            client_release_too_low = 2,
            client_release_too_high = 3,
            invalid_request_operation = 4,
            invalid_request_body = 5,
            invalid_request_body_size = 6,
            session_too_low = 7,
            session_release_mismatch = 8,

            comptime {
                for (std.enums.values(Reason), 0..) |reason, index| {
                    assert(@intFromEnum(reason) == index);
                }
            }
        };
```

Changing release/session is one way to go about it but it involves changing some client-level properties at runtime (which is what we're doing now). However, I'm curious to find out if either of these two would work:
* Sending an _invalid operation_ (perhaps creating a private function within the Rust client that sends a request with an _invalid_ operation number)
* Sending a request with size > `request_size_limit`, which can actually be set as `--request-size-limit` in `tigerbeetle start`

This is where these requests are rejected on the cluster-side!

https://github.com/tigerbeetle/tigerbeetle/blob/f9c4c763962287ec59d5fa4aa112c6a029aae3df/src/vsr/replica.zig#L6179-L6208

Comment (brson) [reply src/clients/c/tb_client/context.zig]: I've looked at the invalid operation option and it works but I am worried that it breaks type safety in multiple ways that are probably UB since the client only deals in operation enums, not raw u8 opcodes. (Could look at how invasive it would be to add an InvalidOpcode variant).

Haven't looked at the large requests yet. Next.

Comment (brson) [reply src/clients/c/tb_client/context.zig]: actually using an invalid operation still requires some invasive changes to the tb_client and doesn't obviously let us remove the additional new testing api; still trying to formulate it in a way that doesn't require adding api surface to tb_client

File: src/clients/c/tb_client/context.zig (hunk 1 above)
Comment (brson) [reply src/clients/c/tb_client/context.zig:218]: just making eviction_reason reason atomic doesn't fix the biggest issue here - even if that check is atomic _both subsequent branches_ still read values that have been deinited already; that is an atomic eviction_reason protects eviction_reason, but not the other stuff in the submit function.

the way to do it would be to take the lock that protects submit any time any of the stuff in submit is used elswhere, but i think it's the wrong approach - requires understanding more synchronization invariants.

i don't see a reason to shut down the io thread before the client owner requests it.

File: src/clients/c/tb_client/context.zig (hunk 2 above)
Comment (brson) [reply src/clients/c/tb_client/context.zig]: Yes I think that would work.

File: src/clients/c/tb_client/context.zig (hunk 3 above)
Comment (brson) [reply src/clients/c/tb_client/context.zig]: Yes that is right. self.pending is only accessed by the io thread so this is safe.

self.submitted is accessed by arbitrary threads and is protected by the interface's lock.

All this should be documented clearly, but I'm waiting until some of this stuff is sorted out and understand it better.

**Comment by brson** [comment]
Status on this is that I have some confidence the fix does fix the eviction crash, but not the windows crash; the test case I added is quite invasive and I am searching for another; and I'm not confident the test actually captures the same crash I've reproduced with other test cases (that are too slow and unreliable to check in), though it does capture _a_ crash that is fixed by the patch.

File: src/clients/c/tb_client/context.zig (hunk 2 above)
Comment (chaitanyabhandari) [reply src/clients/c/tb_client/context.zig]: Ack, perhaps we can take that up as a follow up then... this PR is tricky enough as it is!

File: src/clients/c/tb_client/context.zig (hunk 1 above)
Comment (chaitanyabhandari) [reply src/clients/c/tb_client/context.zig:218]: Ah, that makes sense, we still have the below, and even though `self.submitted.push(packet);` is protected via `client.locker.lock();` so its thread safe, we _could_ still use a deinitialized `self.signal`, since this user thread could race with an deinitializing IO thread...

```C
            self.submitted.push(packet);
            self.signal.notify();
```

Agreed that we needn't shut down the IO thread before the client owner requests it, but perhaps we could still retain the behaviour from https://github.com/tigerbeetle/tigerbeetle/pull/2923 and reclaim client resources immediately as its evicted so the outgoing connections are closed? To that end, I was thinking about whether we could add the client deinitialization to when we're setting `eviction_reason`. IIUC, this is safe to do, as we only access `self.client` from within the `io_thread`, no?

```C
            // After eviction the io_thread will short-circuit packet_send to packet_cancel.
            self.eviction_reason = eviction.header.reason;

            // This is subtle and ugly - client _does not_ send us a completion callback
            // for evicted requests, so if we don't do something with it we'll deadlock.
            // Here we reach into the guts of client and grab the packet to cancel it.
            self.cancel_request_inflight();

            // Cancel every packet.
            while (self.pending.pop()) |packet| {
                packet.assert_phase(.pending);
                self.packet_cancel(packet);
            }
            
            // Deinitialize client instance.
            self.client.deinit(self.gpa.allocator());
            self.message_pool.deinit(self.gpa.allocator());
```

File: src/clients/c/tb_client/context.zig (hunk 4 above)
Comment (chaitanyabhandari) [reply src/clients/c/tb_client/context.zig]: Ah, thanks for trying to explore the invalid operation number route Brian! I don't _fully_ understand the invasive changes to tb_client it entails, but I hope we're able to find a way around it...

Otherwise, perhaps the `--request-size-limit` might be a good option, given we can set it in `tigerbeetle start` in `new_with_name`... and then do a `create_transfers` request that is too big to get the eviction message from the cluster? 🤔

**Review by batiati — COMMENTED** [review]
Here are two high-level considerations:

- Eviction and Shutdown are very similar, and I feel we should try to share the synchronization mechanism already present in `client.close()` for the eviction case, which ultimately is an involuntary close + reason.

- Testability is fundamental, but I don't think we should increase the API surface to introduce a shim for eviction. It belongs to a mocked client/cluster, something that might even replace the EchoClient entirely. My suggestion here would be to do it in a separate PR, although it means applying the fix without a proper testing strategy.

File: src/clients/c/tb_client/context.zig (hunk 4 above)
Comment (batiati) [reply src/clients/c/tb_client/context.zig]: > Otherwise, perhaps the `--request-size-limit` might be a good option, given we can set it in `tigerbeetle start` in `new_with_name`... and then do a `create_transfers` request that is too big to get the eviction message from the cluster? 🤔

We can trigger an eviction by:
- Sending invalid operations or payloads, which is gated by the tb_client implementation. Too intrusive.
- Being at an unsupported version. Extremely intrusive.
- Not having a valid client session! This seems to be the easiest way to trigger an eviction: Just connect N+1 clients to the cluster from another thread so the cluster kicks out the oldest client!

Comment (chaitanyabhandari) [reply src/clients/c/tb_client/context.zig]: Ah, thanks a lot for chiming in @batiati! Indeed the `--request-size-limit` route does not work as well without intrusive changes, since the cluster communicates this limit to the client as a result of registration. So, requests that don't adhere to the cluster-level limit are rejected at the client with `too_much_data`.

https://github.com/tigerbeetle/tigerbeetle/blob/8eff34c11fb0304e5c68ec1d1edb81da93ed3258/src/clients/c/tb_client/context.zig#L735

I like the idea of registering `clients_max` clients from another thread as well, it indeed seems to be the cleanest and least intrusive way!

https://github.com/tigerbeetle/tigerbeetle/blob/8eff34c11fb0304e5c68ec1d1edb81da93ed3258/src/constants.zig#L76-L81

**Comment by brson** [comment]
> Here are two high-level considerations:
> 
>     * Eviction and Shutdown are very similar, and I feel we should try to share the synchronization mechanism already present in `client.close()` for the eviction case, which ultimately is an involuntary close + reason.

I can't find a `close` method on client or context and don't see what synchonization mechanism this refers to.

**Comment by brson** [comment]
I have rebased and pushed just the fix with no tests. This can be merged as-is if desired. But there is no regression test.

The big problem right now is that we just don't have good multithreaded test coverage and its hard to have confidence in any changes here.

**Comment by batiati** [comment]
> I can't find a `close` method on client or context and don't see what synchonization mechanism this refers to.

Sorry Brian, I was thinking of the Java Client! I meant calling the `deinit()` function, which is already synchronized by the client interface. I think it could work (and even simplify the code).

**Comment by brson** [comment]
> > I can't find a `close` method on client or context and don't see what synchonization mechanism this refers to.
> 
> Sorry Brian, I was thinking of the Java Client! I meant calling the `deinit()` function, which is already synchronized by the client interface. I think it could work (and even simplify the code).

I'm afraid I still don't understand. There is a lock in `ClientInterface` that protects the submission queue (and probably the vtable). The prior code that this fixes did seem to be trying to use it to protect the eviction reason and the signal object, but the patch I have made here argues that is the wrong approach and more complicated than what this patch does.

Edit: to be clear, this patch moves the responsibility for dispatching message results for messages that were cancelled because of prior eviction from an arbitrary thread back to the io thread, where all other result dispatches happen. It reduces synchronization requirements and the need to understand them.

**Comment by brson** [comment]
Current test failure doesn't look related.

**Comment by brson** [comment]
I also want to point out one other problem with the current code that I didn't write in the op: because the current code calls (rarely) a user-defined callback on an unexpected thread, this could lead to deadlock:

The user expects callbacks to be called on some other io thread they don't know anything about, but in the one case of submission after eviction we currently call that callback on _the same thread as the submission_. If the user has taken their own lock around the submission and also in the callback it is instant deadlock in a rare case.

**Comment by brson** [comment]
Current ci failure seems to be real: https://github.com/tigerbeetle/tigerbeetle/actions/runs/19877261564/job/56968547845?pr=3373

I have to sleep on this. I am so frustrated.

**Comment by brson** [comment]
Still trying to convince myself this patch isn't going to cause regressions. I am seeing ci failures occassionally but they're timeouts:

```
 +- run scripts failure
2025-12-08 19:44:49.124Z error: process failed with ExecTimeout: mvn --batch-mode --file pom.xml --quiet install
java ci: 14m48.019s
2025-12-08 19:44:49.130Z error: ExecTimeout
/home/runner/work/tigerbeetle/tigerbeetle/src/shell.zig:557:31: 0x1304791 in exec_inner (scripts)
            if (timeout <= 0) return error.ExecTimeout;
                              ^
/home/runner/work/tigerbeetle/tigerbeetle/src/shell.zig:398:5: 0x138266c in exec__anon_38908 (scripts)
    return exec_inner(shell, argv.slice(), .{});
    ^
/home/runner/work/tigerbeetle/tigerbeetle/src/clients/java/ci.zig:21:5: 0x138410d in tests (scripts)
    try shell.exec("mvn --batch-mode --file pom.xml --quiet install", .{});
    ^
/home/runner/work/tiger
```

I suspect this isn't releated to this patch but don't know.

File: src/clients/c/tb_client/context.zig (hunk 4 above)
Comment (brson) [reply src/clients/c/tb_client/context.zig]: > * Not having a valid client session! This seems to be the easiest way to trigger an eviction: Just connect N+1 clients to the cluster from another thread so the cluster kicks out the oldest client!

I've tried this and it isn't currently viable - connecting repeatedly is too slow for unit tests (takes 0.5 s per connection here), and the whole system deadlocks at 32 simultaneous connections - seems to be a separate bug.

**Comment by brson** [comment]
I'm about ready to give up trying to find a reproducible test case and am thinking we should just check this in and hope for the best.

File: src/clients/c/tb_client/context.zig (hunk 1 above)
Comment (brson) [reply src/clients/c/tb_client/context.zig:218]: > Ah, that makes sense, we still have the below, and even though `self.submitted.push(packet);` is protected via `client.locker.lock();` so its thread safe, we _could_ still use a deinitialized `self.signal`, since this user thread could race with an deinitializing IO thread...
> 
> ```c
>             self.submitted.push(packet);
>             self.signal.notify();
> ```
> 
> Agreed that we needn't shut down the IO thread before the client owner requests it, but perhaps we could still retain the behaviour from #2923 and reclaim client resources immediately as its evicted so the outgoing connections are closed? To that end, I was thinking about whether we could add the client deinitialization to when we're setting `eviction_reason`. IIUC, this is safe to do, as we only access `self.client` from within the `io_thread`, no?
> 
> ```c
>             // After eviction the io_thread will short-circuit packet_send to packet_cancel.
>             self.eviction_reason = eviction.header.reason;
> 
>             // This is subtle and ugly - client _does not_ send us a completion callback
>             // for evicted requests, so if we don't do something with it we'll deadlock.
>             // Here we reach into the guts of client and grab the packet to cancel it.
>             self.cancel_request_inflight();
> 
>             // Cancel every packet.
>             while (self.pending.pop()) |packet| {
>                 packet.assert_phase(.pending);
>                 self.packet_cancel(packet);
>             }
>             
>             // Deinitialize client instance.
>             self.client.deinit(self.gpa.allocator());
>             self.message_pool.deinit(self.gpa.allocator());
> ```

Yes, I think shutting down the client while leaving the I/O thread up is a good idea.

Comment (brson) [reply src/clients/c/tb_client/context.zig:218]: I have pushed a commit that makes this change.

**Comment by brson** [comment]
CI failure doesn't seem related.

**Comment by matklad** [comment]
```
Error:  Crashed tests:
Error:  com.tigerbeetle.IntegrationTest
Error:  org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
```

I am not sure it is unrelated, this is how `unreachable`, `assert(false)` and some such in tb_client "C" code manifest on the JVM side.

**Comment by brson** [comment]
Pushing some eviction experiments to run them against ci.

**Comment by brson** [comment]
I've discovered that it is not possible currently to safely deinit the internal client while leaving the io thread running.

This has lead me to look even more suspiciously at `cancel_all`, working on a patch that shuts down pending I/O in a more reliable way, deleting `cancel_all` entirely, and should be more cross-platform compatible. Hopeful that this is a path to solving all client shutdown problems.

**Comment by brson** [comment]
Continued at https://github.com/tigerbeetle/tigerbeetle/pull/3509
